capsules/extra/src/bmi270.rs

```diff
@@ -0,0 +1,520 @@
+                } else if us == 20000 {
+                    buffer[0] = Registers::InternalStatus as u8;
+
+                    if let Err((i2c_err, buffer)) = self.i2c.write_read(buffer, 1, 1) {
+                        self.state.set(State::Sleep);
+                        self.buffer.replace(buffer);
+                        self.ninedof_client
+                            .map(|client| client.callback(i2c_err as usize, 0, 0));
+                    } else {
+                        self.state.set(State::CheckStatus);
+                    }
+                }
+            }
+            _ => {}
+        });
+    }
+}
+
+impl<'a, A: Alarm<'a>, I: I2CDevice> NineDof<'a> for BMI270<'a, A, I> {
+    fn set_client(&self, client: &'a dyn NineDofClient) {
+        self.ninedof_client.set(client);
+    }
+
+    fn read_accelerometer(&self) -> Result<(), ErrorCode> {
+        if !self.pending_accel.get() {
+            self.pending_accel.set(true);
+            if self.pending_gyro.get() {
+                Err(ErrorCode::BUSY)
+            } else {
+                self.start_measurement()
+            }
+        } else {
+            Err(ErrorCode::BUSY)
+        }
+    }
+
+    fn read_gyroscope(&self) -> Result<(), ErrorCode> {
+        if !self.pending_gyro.get() {
+            self.pending_gyro.set(true);
+            if self.pending_accel.get() {
+                Err(ErrorCode::BUSY)
+            } else {
+                self.start_measurement()
+            }
+        } else {
+            Err(ErrorCode::BUSY)
+        }
+    }
+}
+
+impl<'a, A: Alarm<'a>, I: I2CDevice> AlarmClient for BMI270<'a, A, I> {
+    fn alarm(&self) {
+        self.handle_alarm()
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Debug)]
+enum State {
+    Sleep,
+
+    WaitingForAlarm(u32),
+    InitWriteConfig,
+    InitDone,
+
+    CheckStatus,
+
+    ConfAccelRange,
+    ConfGyro,
+    ConfGyroRange,
+    ConfPower,
+    CheckConf,
+    Enable,
+    InitRead,
+    Read,
+    Done,
+    Idle,
+}
+
+impl<'a, A: Alarm<'a>, I: I2CDevice> I2CClient for BMI270<'a, A, I> {
+    fn command_complete(&self, buffer: &'static mut [u8], status: Result<(), i2c::Error>) {
+        if let Err(i2c_err) = status {
+            self.state.set(State::Sleep);
+            self.buffer.replace(buffer);
+            self.ninedof_client
+                .map(|client| client.callback(i2c_err as usize, 0, 0));
+            return;
+        }
+
+        match self.state.get() {
+            State::WaitingForAlarm(us) => {
+                self.buffer.replace(buffer);
+                let delay = self.alarm.ticks_from_us(us);
+                self.alarm.set_alarm(self.alarm.now(), delay);
+            }
+            State::InitWriteConfig => {
+                self.config_file.take().map(|config_file| {
+                    if let Err((i2c_err, buffer)) = self.i2c.write(config_file, 8193) {
+                        self.state.set(State::Sleep);
+                        self.buffer.replace(buffer);
```

Comment:
I believe this should be replacing the `self.config_file` and not the buffer, right?